topics/basics/testing_plugins/integration_tests/integration_tests_ui.md

```diff
@@ -0,0 +1,198 @@
+
+The element corresponds to the following button:
+
+![](integration_tests_ui_sample.png){width=706}
+
+Similar to web testing frameworks like Selenium, XPath is used to locate components.
+The Driver framework provides a simple XPath builder.
+Here are several ways to find the same component:
+
+```kotlin
+xQuery { byVisibleText("Current File") }
+xQuery { byAccessibleName("Current File") }
+xQuery { byType("com.intellij.execution.ui.RedesignedRunConfigurationSelector\$createCustomComponent$1") }
+```
+
+For reliable component identification, prioritize these attributes: `accessiblename`, `visible_text`, `icon`, `javaclass`.
+Multiple attributes can be combined for a more precise selection.
+
+## Interaction with Components
+
+Once a component is located, it's possible to interact with it or verify its properties.
+
+To click the <control>Current File</control> button:
+
+```kotlin
+x(xQuery { byVisibleText("Current File") }).click()
+```
+
+The `x()` call creates a lazy reference to the component.
+It means that the XPath query isn't executed immediately and component lookup happens only when an action (like `click()`) is invoked.
+
+Here's a part of test that incorporates UI interaction:
+```kotlin
+runIdeWithDriver().useDriverAndCloseIde {
+  waitForIndicators(1.minutes)
+  ideFrame {
+    x(xQuery { byVisibleText("Current File") }).click()
+  }
+}
+```
+
+Beyond mouse clicks, keyboard input and shortcuts can be simulated:
+
+```kotlin
+keyboard {
+  enterText("Sample text")
+  enter()
+  hotKey(if (SystemInfo.isMac) KeyEvent.VK_META else KeyEvent.VK_CONTROL, KeyEvent.VK_A)
+  backspace()
+}
+```
+
+> On macOS, the interaction via `java.awt.Robot` requires special permissions.
+> IntelliJ IDEA should be granted the necessary permissions via the Accessibility page, which can be found under _System Settings | Privacy & Security_.
+>
+{style="note"}
+
+## Asserting Properties
+
+The complete UI test:
+
+```kotlin
+@Test
+fun simpleTestForCustomUIElement() {
+  Starter.newContext(
+    "testExample",
+    TestCase(
+    IdeProductProvider.IC,
+    GitHubProject.fromGithub(branchName = "master",
+      repoRelativeUrl = "JetBrains/ij-perf-report-aggregator"))
+      .withVersion("2024.3")
+    ).apply {
+  val pathToPlugin = System.getProperty("path.to.build.plugin")
+  PluginConfigurator(this).installPluginFromFolder(File(pathToPlugin))
+  }.runIdeWithDriver().useDriverAndCloseIde {
+      waitForIndicators(1.minutes)
+      ideFrame {
+      x(xQuery { byVisibleText("Current File") }).click()
+      val configurations = popup().jBlist(xQuery { contains(byVisibleText("Edit Configurations")) })
+      configurations.shouldBe("Configuration list is not present", present)
+      Assertions.assertTrue(configurations.rawItems.contains("backup-data"),
+        "Configurations list doesn't contain 'backup-data' item: ${configurations.rawItems}")
+    }
+  }
+}
+```
+
+The test does the following:
+
+* Opening the popup
```

Comment:
Looking at the test, clicking the button is first, so I don't understand what popup is opened in the the beggining.

I think it would be better to add `// 1`, `// 2`, etc. comment in the code snippet and reference them below with a numbered list.